macaw-aarch32/src/Data/Macaw/ARM/Identify.hs

```diff
@@ -74,3 +90,122 @@ isReturnValue absProcState val =
+-- more general infrastructure.
+simplifiedMux
+  :: MSS.SimplifierExtension arch
+  => MC.Value arch ids tp
+  -> Maybe (MC.App (MC.Value arch ids) tp)
+simplifiedMux ipVal
+  | Just app@(MC.Mux {}) <- MC.valueAsApp ipVal =
+      MSS.simplifyArchApp app <|> pure app
+  | otherwise = Nothing
+
+data ReturnsOnBranch = ReturnsOnTrue | ReturnsOnFalse
+  deriving (Eq)
+
+-- | Inspect the IP register to determine if this statement causes a conditional return
+--
+-- We expect a mux where one of the IP values is the abstract return address and
+-- the other is an executable address.  Ideally we would be able to say that is
+-- the "next" instruction address, but we do not have a good way to determine
+-- the *current* instruction address. This just means that we have a more
+-- flexible recognizer for conditional returns, even though there are
+-- (probably?) no ARM instructions that could return that way.
+--
+-- The returned values are:
+--
+--  * The condition of the conditional return
+--  * The next IP
+--  * An indicator of which branch is the return branch
+--  * The statements to use as the statement list
+--
+-- Note that we currently don't modify the statement list, but could
+identifyConditionalReturn
+  :: MC.Memory 32
+  -> Seq.Seq (MC.Stmt ARM.AArch32 ids)
+  -> MC.RegState (MC.ArchReg ARM.AArch32) (MC.Value ARM.AArch32 ids)
+  -> MA.AbsProcessorState (MC.ArchReg ARM.AArch32) ids
+  -> Maybe ( MC.Value ARM.AArch32 ids MT.BoolType
+           , MC.ArchSegmentOff ARM.AArch32
+           , ReturnsOnBranch
+           , Seq.Seq (MC.Stmt ARM.AArch32 ids)
+           )
+identifyConditionalReturn mem stmts s finalRegState
+  | not (null stmts)
+  , Just (MC.Mux _ c t f) <- simplifiedMux (s ^. MC.boundValue MC.ip_reg) =
+      case asReturnAddrAndConstant mem finalRegState t f of
+        Just nextIP -> return (c, nextIP, ReturnsOnTrue, stmts)
+        Nothing -> do
+          nextIP <- asReturnAddrAndConstant mem finalRegState f t
+          return (c, nextIP, ReturnsOnFalse, stmts)
+  | otherwise = Nothing
+
+-- | Recognize ARM conditional returns and generate an appropriate arch-specific
+-- terminator
+--
+-- Conditional returns are not supported in core macaw, so we need to use an
+-- arch-specific terminator.  Unlike simple arch-terminators, this one requires
+-- analysis that can only happen in the context of a block classifier.
+--
+-- Note that there are two cases below that could be handled. It seems unlikely
+-- that these would be produced in practice, so they are unhandled for now.
+conditionalReturnClassifier :: MAI.BlockClassifier ARM.AArch32 ids
+conditionalReturnClassifier = do
+  stmts <- CMR.asks MAI.classifierStmts
+  mem <- CMR.asks (MAI.pctxMemory . MAI.classifierParseContext)
+  regs <- CMR.asks MAI.classifierFinalRegState
+  absState <- CMR.asks MAI.classifierAbsState
+  Just (cond, nextIP, returnBranch, stmts') <- return (identifyConditionalReturn mem stmts regs absState)
```

Comment:
Should the `fail` path here invoke (i.e. fall through to) the `defaultClassifier`?

Reply:
Yeah a pattern match failure will fall through (and its error message will be collected). I'll try to restructure so that the more precise messages are preserved.